plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusingConnectionFactory.java

```diff
@@ -0,0 +1,205 @@
+{
+    @GuardedBy("this")
+    private final Map<String, Connection> connections;
+    private final ConnectionFactory delegate;
+
+    @Inject
+    public ReusingConnectionFactory(@ForLazyConnectionFactory ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(2), 10);
+    }
+
+    ReusingConnectionFactory(ConnectionFactory delegate, Duration duration, long maximumSize)
+    {
+        requireNonNull(duration, "duration is null");
+        NonLoadableCache<String, Connection> connections = SafeCaches.buildNonLoadableCache(CacheBuilder.newBuilder()
+                .maximumSize(maximumSize)
+                .expireAfterWrite(duration)
+                .removalListener(notification -> {
+                    if (notification.getCause() == EXPLICIT) {
+                        // connection was taken from the cache
+                        return;
+                    }
+                    try {
+                        requireNonNull(notification.getValue(), "notification.getValue() is null");
+                        notification.getValue().close();
+                    }
+                    catch (SQLException e) {
+                        throw new TrinoException(JDBC_ERROR, e);
+                    }
+                }));
+        this.connections = connections.asMap();
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.remove(queryId);
+        if (connection != null) {
+            return connection;
+        }
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public void beginQuery(ConnectorSession session) {}
+
+    @Override
+    public void cleanupQuery(ConnectorSession session)
+    {
+        Connection connection = connections.remove(session.getQueryId());
+        if (connection != null) {
+            try {
+                connection.close();
+            }
+            catch (SQLException e) {
+                throw new TrinoException(JDBC_ERROR, e);
+            }
+        }
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        for (Connection connection : connections.values()) {
+            connection.close();
+        }
+        connections.clear();
+        delegate.close();
+    }
+
+    private final class CachedConnection
+            extends ForwardingConnection
+    {
+        private final String queryId;
+        private final Connection delegate;
+        private boolean closed;
+        private boolean dirty;
```

Comment:
We kind of sure, that connection start and finish in the same thread, and don't need to Guard this?